server_v2.py

```diff
@@ -229,17 +242,24 @@ def execute_stmt(conn, sqls, stmt):
     try:
         changes_before = conn.total_changes()
         prepared, sql_rest = conn.prepare(sql)
+        if not prepared:
+            raise ResponseError(f"Could not prepare statement: {sql}", "ARGS_INVALID")
+
         param_count = prepared.param_count()
 
         if len(sql_rest.strip()) != 0:
             raise ResponseError(f"SQL string contains more than one statement")
 
-        for param_i, arg_value in enumerate(stmt.get("args", []), 1):
-            if param_i > param_count:
-                raise ResponseError(f"Statement accepts only {param_count} params")
+        args = stmt.get("args", [])
+        named_args = stmt.get("named_args", [])
+        provided_params_count = len(args) + len(named_args)
+        if provided_params_count != param_count:
+            raise ResponseError(f"Required {param_count} but {provided_params_count} were provided", "ARGS_INVALID")
```

Comment:
This check is unnecessarily strict. SQLite does not need you to specify values for all parameters, they will default to NULL (so `provided_arg_count < param_count`).

It might also sometimes be necessary for named parameters to overwrite positional parameters. For example, in the query `SELECT ?1, :two, ?3`, you need to pass `args = ["value for ?1", None, "value ?3"]` and `named_args = {":two": "value for :two"}` if you want to pass the value of `:two` by name (so `provided_arg_count > param_count`).